File: First_Year/1.2_Python/Queue.py
class Queue:
    def __init__(self):
        self.queue = []

    def enqueue(self, item):
        """Add an item to the end of the queue."""
        self.queue.append(item)

    def dequeue(self):
        """Remove and return the front item from the queue. Raises an exception if the queue is empty."""
        if self.is_empty():
            raise IndexError("Dequeue from an empty queue")
        return self.queue.pop(0)

    def front(self):
        """Return the front item of the queue without removing it. Raises an exception if the queue is empty."""
        if self.is_empty():
            raise IndexError("Front from an empty queue")
        return self.queue[0]

    def is_empty(self):
        """Return True if the queue is empty, False otherwise."""
        return len(self.queue) == 0

    def size(self):
        """Return the number of items in the queue."""
        return len(self.queue)

# Create a queue
q = Queue()

# Enqueue elements into the queue
q.enqueue(14)
q.enqueue(3)
q.enqueue(5)

# Get the front element
print("Front element is:", q.front())  # Output: Front element is: 14

# Dequeue the front element
print("Dequeued element is:", q.dequeue())  # Output: Dequeued element is: 14

# Check if the queue is empty
print("Is queue empty?", q.is_empty())  # Output: Is queue empty? False

# Get the current size of the queue
print("Current size of the queue:", q.size())  # Output: Current size of the queue: 2

File: 2.2 CPP_DSA/Queue.cpp
#include <iostream>
using namespace std;

#define SIZE 5 // Define the maximum size of the queue

class Queue {
private:
    int items[SIZE]; // Array to hold the queue elements
    int front, rear; // Front and rear indices

public:
    Queue() { // Constructor to initialize the queue
        front = -1;
        rear = -1;
    }

    // Function to check if the queue is full
    bool isFull() {
        return (rear == SIZE - 1);
    }

    // Function to check if the queue is empty
    bool isEmpty() {
        return (front == -1 || front > rear);
    }

    // Function to add an element to the queue
    void enqueue(int element) {
        if (isFull()) {
            cout << "Queue is full!" << endl;
            return;
        }
        if (front == -1) { // First element being added
            front = 0;
        }
        rear++;
        items[rear] = element; // Add the new element
        cout << "Inserted " << element << endl;
    }

    // Function to remove an element from the queue
    int dequeue() {
        if (isEmpty()) {
            cout << "Queue is empty!" << endl;
            return -1; // Return an error value
        }
        int removedElement = items[front];
        front++;
        return removedElement; // Return the removed element
    }

    // Function to get the front element of the queue
    int peek() {
        if (isEmpty()) {
            cout << "Queue is empty!" << endl;
            return -1; // Return an error value
        }
        return items[front]; // Return the front element
    }
};

// Example usage of the Queue class
int main() {
    Queue q;

    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);
    
    cout << "Front element is: " << q.peek() << endl;

    cout << "Dequeued: " << q.dequeue() << endl;
    
    cout << "Front element after dequeue: " << q.peek() << endl;

    while (!q.isEmpty()) {
        cout << "Dequeued: " << q.dequeue() << endl;
    }

    q.dequeue(); // Attempt to dequeue from an empty queue

    return 0;
}
